### PR TITLE
Add stackoverflow-like tags #1003

### LINKED ISSUE #1169: Docker-Compose 3.0 "Extends" functionality removed: "docker-compose run" won't work until rewritten without using extends
Current version of docker-compose.yml in use makes load-bearing use of extends. Support for extends was removed as of docker-compose version 3, our compose file is version 3.4. 

I believe this is why docker-compose run... is failing.

See "Removed" from Version 3: https://github.com/docker/docker.github.io/blob/master/compose/compose-file/compose-versioning.md#version-3

Unfortunately, the reference specifications for compose to not make a note that extends was deprecated as of 3.0 (https://github.com/docker/docker.github.io/blob/master/compose/compose-file/compose-versioning.md).

### LINKED ISSUE #1182: Offering and Looking for fields aren't highlighted in the search results
add `ais-snippet` tag

### LINKED ISSUE #1185: Join profile edit pages into one
### Current state  
We have 4 different profile edit pages
- General  
- Community
- Career
- Causes Area

This is suboptimal UX because it means users have to click a lot to change things  

### Desired state
All edit pages on 1 page: /profile/edit so that users have only 1 place to go to  

### Acceptance criteria
- [ ] All profile edit fields on 1 page: /profile/edit
- [ ] Visually Separated into subsection corresponding to the previous one (General, Community, Career, cause area)
- [ ] All other edit pages and URLs removed (profile/edit_profile_career, etc)  
- [ ] Link in verification email should redirect to /profile/edit to nudge people to provide more info (in lieu of depracated "first_visit" url)
- [ ] Remove "first_visit" URL and related message

### LINKED ISSUE #1190: When selecting organisers when editing group, unapproved profiles are shown
image

We should filter unapproved users from this list.

### PR DESCRIPTION
**Pull request by viktor-yunenko**
- related issue #1003 
- dev server - https://eahub-tags-d01de1e.us.aldryn.io/

Before release:  
- [ ] provide an admin URL with the names that had more than 3 names
- [x] test empty tags creation
- [x] add cause_areas_other and `other` to the search
- [x] add fields for cause areas expertise and affiliations
- [x] move tags to the bottom of profile edit list
- [x] check why cause areas contain expertise
- [x] show "topics I speak about" as a tag field
- [x] make a decision on whether to allow creating cause area tags @vagarwalla  

Testing:
- [x] check whether local groups need bootstrap-multiselect
- [x] forms error messages render
- [x] tags & local groups profile logging
- [x] multi browser testing

Optional or future tickets:
- [ ] Other field: toggle on profile edit page for users who have it filled out and those who haven't  
- [x] Add 2-3 sections to profile edit page to make it less cluttered. [Ideas for sections](https://docs.google.com/spreadsheets/d/1Azn3CJ5HDk1PmO3o7olbvI9t8s0NqLvqkqZ-g-rKjmY/edit?usp=sharing_eip&ts=605725e1)  

Related tickets:
- closes #1169
- closes #1185
- closes #1190
- closes #1182

Release aim: Friday, 2nd of April 2021

### CODE REVIEW

File: eahub/base/static/components/mailchimp/mailchimp.js
Comment (sebgrebe) [inline eahub/base/static/components/mailchimp/mailchimp.js:1]: What was the mailchimp code doing and why are we removing it?

File: eahub/base/static/components/profile-edit-image.js
Comment (sebgrebe) [inline eahub/base/static/components/profile-edit-image.js:1]: I assume it's tricky to separate the multiselect stuff out from other profile edit actions? If not, it would be nice to move the non-tag related JS (i.e., non multiselect stuff) at a later stage to reduce the scope of this MR.

If that's too difficult, we need a full QA list of all the things we used to be able to do on editing profile to make sure we're not losing any functionality when migrating from vanilla to vue JS on this.

File: eahub/config/settings.py
Comment (sebgrebe) [inline eahub/config/settings.py:73]: Why this change?

File: eahub/config/settings.py
Comment (sebgrebe) [inline eahub/config/settings.py]: Why are we adding RECAPTCHA and DOMAIN here?

File: eahub/profiles/api/serializers.py
Comment (sebgrebe) [inline eahub/profiles/api/serializers.py]: Probably to be discussed with product, but what's the purpose of the generic tag?

File: eahub/profiles/api/views.py
Comment (sebgrebe) [inline eahub/profiles/api/views.py]: Is this about the "merge" actions we talked? So do you want duplicates to be handled code or rather by admins?

File: eahub/profiles/management/commands/profilesfromlegacyhub.py
Comment (sebgrebe) [inline eahub/profiles/management/commands/profilesfromlegacyhub.py]: I wonder whether we can get rid of this code. The migration from legacy hub has happened, not sure it's worth our effort to maintain this code

File: eahub/profiles/tests/test_tags_api.py
Comment (sebgrebe) [inline eahub/profiles/tests/test_tags_api.py:62]: Is this test dependent on test_addition having run first? And is this test order guaranteed?

**Review by sebgrebe — CHANGES_REQUESTED** [review]
Initial review with clarificatory questions. Once we have speced this out I can do another review (also once  the bs5 PR is in, bit tricky to review this as the changes are in here too)

File: eahub/config/settings.py
Comment (viktor-yunenko) [reply eahub/config/settings.py:73]: "FeaturePolicyMiddleware" has been depracated 1-2 years ago, it's now called "permissions"

File: eahub/profiles/api/views.py
Comment (viktor-yunenko) [reply eahub/profiles/api/views.py]: here I would need to add better error handling for the sake of UX, for now the user would just be unable to add the tag and wouldn't see any error message, which would be frustrating

File: eahub/profiles/management/commands/profilesfromlegacyhub.py
Comment (viktor-yunenko) [reply eahub/profiles/management/commands/profilesfromlegacyhub.py]: I agree here, I've been thinking about scheduling a 1-2 week-long cleanup once we're done with the pending tasks and have a bit of time. Ie to follow the best practice of refactoring sprints. 

I would say the only reason I haven't been advocating for one is that we're still in the project feasibility assessment stage.

File: eahub/profiles/tests/test_tags_api.py
Comment (viktor-yunenko) [reply eahub/profiles/tests/test_tags_api.py:62]: not really - django runs each test in a db transaction, hence here django regenerates the tags for every test function

File: eahub/config/settings.py
Comment (viktor-yunenko) [reply eahub/config/settings.py]: they're needed in few templates. In short it's perfectly ok to add to this list everything, except the secrets as `RECAPTCHA_SECRET_KEY` which must never be rendered through HTML

File: eahub/base/static/components/mailchimp/mailchimp.js
Comment (viktor-yunenko) [reply eahub/base/static/components/mailchimp/mailchimp.js:1]: I have no idea really, we've quite a bit of dead code around, afaik this is one of those pieces.

File: eahub/base/static/components/profile-edit-image.js
Comment (viktor-yunenko) [reply eahub/base/static/components/profile-edit-image.js:1]: I agree, I would be happy to look into it during the profile editing form composition

File: eahub/profiles/api/serializers.py
Comment (viktor-yunenko) [reply eahub/profiles/api/serializers.py]: the idea of mine was to unblock users - eg if somebody thinking that `couchsurfing` tag is a must, now instead of squeezing it into something nonsensual as "expertise area" they can add a tag as `Helping with couchsurfing" or "Looking for couchsurfing". After that we can notice those tags, analyze the demand and consider creating a better organized field or a set of fields.

I would say if we were god-like human beings who knew everything about our users' desires we wouldn't need a field as `tags_generic` :)

**Comment by sebgrebe** [comment]
[tasks moved to the PR description]

File: eahub/base/static/components/profile-edit-image.js
Comment (viktor-yunenko) [reply eahub/base/static/components/profile-edit-image.js:1]: all logic form this file was replaced with django-select2 and native bootstrap5 styles

File: eahub/profiles/api/serializers.py
Comment (viktor-yunenko) [reply eahub/profiles/api/serializers.py]: I'm planning to import the values from EAG "Affinity" field, eg "People managers", "Earning to give", "Parenting", etc

**Comment by vagarwalla** [comment]
Cause area tags

Can we allow people to suggest them but it isn't visible unless an admin approves it?
Otherwise, do not allow for now. Maybe other causes can go in the general tag box later one.

Reasoning: 
- If we tried to integrate that data with CEA's data in the future for any reason, it would probably be messy

**Comment by viktor-yunenko** [comment]
As I mentioned before I'm neutral about preventing the creation of cause areas tag.

I wouldn't like to base our decisions on the fields sharing though, mostly after working with the data that we've got from EAG.

Getting the data is amazing btw, thank you Vaidehi a lot for organizing it!

We've been maintaining the fields to correspond to what we thought CEA was collecting, but in the latest export after I exclude the generic fields (email, name, country/city), we're left with only one kind-of-compatible-field - Local Groups.

One can say perhaps two compatible fields, if we're willing to assume that `Area of expertise/study` is directly importable, because it doesn't fit naturally into only `Expertise areas` or `Career interest`.

In short I don't expect EA organizations to have a stable idea of what fields they want to have, most likely it's going to change by the time of our next import.

**Comment by viktor-yunenko** [comment]
> If we tried to integrate that data with CEA's data in the future for any reason, it would probably be messy

Hence yes, right now it's as messy as it can get :)

Not a significant issue for me though, as a data import isn't a big deal. Fixing UX problems caused by pseudo-compatibility could be a major issue for us though.

**Comment by viktor-yunenko** [comment]
Hi @sebgrebe, I didn't test the logging (which probably doesn't work) and will be doing eg multi-browser testing later. But you can already start reviewing the code. Would you like to do it over 1-2 calls? I can be around eg on Saturday.

File: eahub/localgroups/forms.py
Comment (viktor-yunenko) [inline eahub/localgroups/forms.py:29]: I wasn't sure why "superuser" here is important, hence I removed it for simplicity sake, otherwise it's hard to test since I'm seeing bots in the dropdown as an admin. I recommend to do anything admin related in the admin panel.

File: eahub/base/static/components/profile/edit/profile-edit.ts
Comment (viktor-yunenko) [inline eahub/base/static/components/profile/edit/profile-edit.ts]: I added it to the todo

**Comment by viktor-yunenko** [comment]
I also added the logs for tags and local groups.

File: docker-compose.yml
Comment (sebgrebe) [inline docker-compose.yml:1]: why?

File: eahub/templates/eahub/edit_profile.html
Comment (sebgrebe) [inline eahub/templates/eahub/edit_profile.html]: Tag "Topics I speak about" missing

File: eahub/profiles/migrations/0022_add_field_last_name.py
Comment (sebgrebe) [inline eahub/profiles/migrations/0022_add_field_last_name.py:56]: I'd add "Public profiles are visible on the web."

File: eahub/templates/eahub/edit_profile.html
Comment (sebgrebe) [inline eahub/templates/eahub/edit_profile.html]: Maybe rename to "Other tags"? @vagarwalla, any native speaker suggestions? We currently use "Tags" for all tags we haven't assigned to categories like "Cause Area", i.e., it's a dustbin category. Other idea would be "Miscellaneous"

File: eahub/localgroups/tests/test_localgroups_models.py
Comment (sebgrebe) [inline eahub/localgroups/tests/test_localgroups_models.py]: was this removed accidentally?

File: eahub/templates/eahub/edit_profile.html
Comment (sebgrebe) [inline eahub/templates/eahub/edit_profile.html:203]: add note that email won't be visible to other users

**Review by sebgrebe — CHANGES_REQUESTED** [review]

File: docker-compose.yml
Comment (sebgrebe) [reply docker-compose.yml:1]: Answered: Because we're actually using version 2

File: eahub/templates/account/signup.html
Comment (sebgrebe) [inline eahub/templates/account/signup.html:51]: Needs to be cleaned up: 
image

Placeholders "FullName" are confusing

Comment (sebgrebe) [reply eahub/templates/account/signup.html:51]: Same for login (I did a hard refresh)
image

File: eahub/templates/eahub/edit_profile.html
Comment (sebgrebe) [inline eahub/templates/eahub/edit_profile.html:170]: On QA session with @vagarwalla we noticed that there are some non-cause areas listed here, like "Software Engineering" and "Giving What We Can"

File: eahub/localgroups/tests/test_localgroups_models.py
Comment (viktor-yunenko) [reply eahub/localgroups/tests/test_localgroups_models.py]: restored 👌

File: eahub/templates/account/signup.html
Comment (viktor-yunenko) [reply eahub/templates/account/signup.html:51]: a caching issue

Comment (sebgrebe) [reply eahub/templates/account/signup.html:51]: Turned out I had to rerun ```npm ci``` because of new dependencies, so it served the old build

File: eahub/templates/eahub/edit_profile.html
Comment (sebgrebe) [inline eahub/templates/eahub/edit_profile.html]: I think it's standard to have city/town before country

File: eahub/templates/eahub/edit_profile.html
Comment (sebgrebe) [inline eahub/templates/eahub/edit_profile.html:109]: image
This sounds to me a bit like "Never create a tag". Suggestion: "Try to find an existing tag before creating a new one"

File: eahub/templates/eahub/edit_profile.html
Comment (sebgrebe) [inline eahub/templates/eahub/edit_profile.html:110]: image

EAHub is missing :) @vagarwalla could you do a systematic check on your QA whether we still have all the items on each list (cause area, career interest, skills, organisational affiliation) we used to have?

File: eahub/templates/eahub/edit_profile.html
Comment (sebgrebe) [inline eahub/templates/eahub/edit_profile.html:238]: Footer missing

image

File: eahub/templates/eahub/edit_profile.html
Comment (sebgrebe) [inline eahub/templates/eahub/edit_profile.html:130]: I only get this locally, not on the deployed server: On selecting an event, this occurs:
image

Any idea how this could be fixed? Not urgent prior to release, but ideally we have the same functionality locally

File: eahub/templates/eahub/profile.html
Comment (sebgrebe) [inline eahub/templates/eahub/profile.html:128]: offering and looking for not on same height

File: eahub/templates/eahub/profile.html
Comment (sebgrebe) [inline eahub/templates/eahub/profile.html:135]: (optional) I think Edit button and social media buttons should be at same height, looks a bit wonky.
If you don't want to do this now, you could add this to some list for to be done later

image

File: eahub/profiles/models.py
Comment (sebgrebe) [inline eahub/profiles/models.py:693]: Export on profiles misses a number of fields, among others:
- Organisational affiliation
- speech topic
- open to volunteer
- social media pages

The custom code you removed was relying on this method here to get all those fields, and the 3rd party library we're using now seems to be missing them.

**Review by sebgrebe — CHANGES_REQUESTED** [review]

File: eahub/profiles/models.py
Comment (viktor-yunenko) [reply eahub/profiles/models.py:693]: I know, we don't need to handle those fields in a custom manner anymore, as they're handled automatically when added to the `fields = []` attribute of the resource class.

I added few more fields that you mentioned, if you want more you can add them as well.

Does that sound sufficient?

File: eahub/templates/eahub/edit_profile.html
Comment (viktor-yunenko) [reply eahub/templates/eahub/edit_profile.html:130]: Did you clone the database from the tags branch before? This list comes from algolia index from the tags db, it won't work without the identical SQL rows.

File: eahub/templates/eahub/profile.html
Comment (viktor-yunenko) [reply eahub/templates/eahub/profile.html:128]: fixed

File: eahub/templates/eahub/edit_profile.html
Comment (viktor-yunenko) [reply eahub/templates/eahub/edit_profile.html]: I think it was standard with paper latters to make it easier for the postman to shuffle through those? which doesn't sound as an appealing standard base to me

I can change it if you're sure

File: eahub/templates/eahub/profile.html
Comment (viktor-yunenko) [reply eahub/templates/eahub/profile.html:135]: yes, it's a firefox bug

FYI I develop only on chrome, hence until multi-browser testing is done I don't look at other browser

File: eahub/templates/eahub/edit_profile.html
Comment (viktor-yunenko) [reply eahub/templates/eahub/edit_profile.html:170]: moved up to the todo list in the ticket

File: eahub/profiles/models.py
Comment (sebgrebe) [reply eahub/profiles/models.py:693]: Fine for now, when we want to export data to other parties we probably want all fields, but we can change that then

File: eahub/base/static/components/profile/edit/profile-tag-input.ts
Comment (sebgrebe) [inline eahub/base/static/components/profile/edit/profile-tag-input.ts]: optional: add error message to alert

File: eahub/localgroups/forms.py
Comment (sebgrebe) [reply eahub/localgroups/forms.py:29]: Nice, this seems to have fixed this issue! #1190

File: eahub/templates/eahub/edit_profile.html
Comment (sebgrebe) [reply eahub/templates/eahub/edit_profile.html]: Had to look this up, it seems that what's conventional differs on world regions: https://ux.shopify.com/designing-address-forms-for-everyone-everywhere-f481f6baf513

As I'm used to the western system, I found the order you used counterintuitive, but other users will differ.

I would suggest though that we should keep it aligned with the groups editing form. So either we need to swap it around there, too (country, city) or we put it back to the original order here (city, country). Both works for me.

File: eahub/templates/eahub/talentsearch.html
Comment (sebgrebe) [inline eahub/templates/eahub/talentsearch.html]: We should remove this page altogether! We forgot to remove it when we introduced the new profiles search:
- https://eahub.org/volunteers/  
- https://eahub.org/speakers/  
- https://eahub.org/candidates   

Can be done in a separate MR (but let's create a ticket to not forget)

**Review by sebgrebe — CHANGES_REQUESTED** [review]

File: eahub/templates/eahub/talentsearch.html
Comment (viktor-yunenko) [reply eahub/templates/eahub/talentsearch.html]: always happy to drop dead code - done 👍

**Comment by vagarwalla** [comment]
1. Cause Area Mapping done here: https://docs.google.com/spreadsheets/d/1Y_GKftJeWh5LBYflnPJwrFrLN6BgIiTGzM-CA-fGK8I/edit#gid=189592928
2. Tag Field Descriptions: in the same doc in a different tab.

File: eahub/templates/eahub/edit_profile.html
Comment (viktor-yunenko) [reply eahub/templates/eahub/edit_profile.html:238]: for ux reasons as discussed

**Comment by viktor-yunenko** [comment]
I moved the description discussion to a google doc to simplify commenting - https://docs.google.com/document/d/1KnoedOwlJGqdzHSkCtqhJpBJX-WKUcUPUWUhU1leUaE/edit

File: eahub/templates/eahub/edit_profile.html
Comment (viktor-yunenko) [reply eahub/templates/eahub/edit_profile.html:109]: done

File: eahub/profiles/migrations/0022_add_field_last_name.py
Comment (viktor-yunenko) [reply eahub/profiles/migrations/0022_add_field_last_name.py:56]: done

**Comment by viktor-yunenko** [comment]
I finished processing the changes. @sebgrebe, please feel free to check. I'll proceed with the deployment at 11-12 by GMT if nobody minds.

**Comment by viktor-yunenko** [comment]
We might want to discuss labels in detail based on my google doc comments, but I'm assuming those aren't blockers?

File: eahub/templates/eahub/edit_profile.html
Comment (sebgrebe) [reply eahub/templates/eahub/edit_profile.html:110]: this still seems to be missing, @victor-yunenko? And did we do a check that all tags have been successully moved? Wondering because if EAHub is missing, other things might be missing, too. Or does the tags server not have the complete db?

Comment (viktor-yunenko) [reply eahub/templates/eahub/edit_profile.html:110]: afaik it's called Effective Altruism Hub

but we of course should add synonyms as "eahub"

Comment (viktor-yunenko) [reply eahub/templates/eahub/edit_profile.html:110]: and yes I've been testing the enums migration few times, as I needed to reset the database and rerun the migrations every week or something

so far the only issue we had is that I thought that "cause areas" are the same as "expertise areas", but it was just a silly assumption rather than a real bug

**Pull request merged**
